pkg/awsutils/awsutils.go

```diff
@@ -819,52 +821,98 @@ func (cache *EC2InstanceMetadataCache) createENI(useCustomCfg bool, sg []*string
-			SubnetId:                       aws.String(cache.subnetID),
-			TagSpecifications:              tagSpec,
-			SecondaryPrivateIpAddressCount: aws.Int64(int64(needIPs)),
+		if !goodTag {
+			continue
 		}
-	}
-	if useCustomCfg {
-		log.Info("Using a custom network config for the new ENI")
-		if len(sg) != 0 {
-			input.Groups = sg
+
+		input := &ec2.CreateNetworkInterfaceInput{}
+
+		if cache.enablePrefixDelegation {
+			input = &ec2.CreateNetworkInterfaceInput{
+				Description:       aws.String(eniDescription),
+				Groups:            aws.StringSlice(cache.securityGroups.SortedList()),
+				SubnetId:          subnet.SubnetId,
+				TagSpecifications: tagSpec,
+				Ipv4PrefixCount:   aws.Int64(int64(needIPs)),
+			}
 		} else {
-			log.Warnf("No custom networking security group found, will use the node's primary ENI's SG: %v", aws.StringValueSlice(input.Groups))
+			input = &ec2.CreateNetworkInterfaceInput{
+				Description:                    aws.String(eniDescription),
+				Groups:                         aws.StringSlice(cache.securityGroups.SortedList()),
+				SubnetId:                       subnet.SubnetId,
+				TagSpecifications:              tagSpec,
+				SecondaryPrivateIpAddressCount: aws.Int64(int64(needIPs)),
+			}
+		}
+
+		if useCustomCfg {
+			log.Info("Using a custom network config for the new ENI")
+			if len(sg) != 0 {
+				input.Groups = sg
+			} else {
+				log.Warnf("No custom networking security group found, will use the node's primary ENI's SG: %v", aws.StringValueSlice(input.Groups))
+			}
+			input.SubnetId = aws.String(customSubnet)
+		} else {
+			log.Info("Using same config as the primary interface for the new ENI")
 		}
-		input.SubnetId = aws.String(subnet)
-	} else {
-		log.Info("Using same config as the primary interface for the new ENI")
-	}
 
-	log.Infof("Creating ENI with security groups: %v in subnet: %s", aws.StringValueSlice(input.Groups), aws.StringValue(input.SubnetId))
+		log.Infof("Creating ENI with security groups: %v in subnet: %s", aws.StringValueSlice(input.Groups), aws.StringValue(input.SubnetId))
 
-	start := time.Now()
-	result, err := cache.ec2SVC.CreateNetworkInterfaceWithContext(context.Background(), input)
-	prometheusmetrics.Ec2ApiReq.WithLabelValues("CreateNetworkInterface").Inc()
-	prometheusmetrics.AwsAPILatency.WithLabelValues("CreateNetworkInterface", fmt.Sprint(err != nil), awsReqStatus(err)).Observe(msSince(start))
-	if err != nil {
+		start = time.Now()
+		result, err = cache.ec2SVC.CreateNetworkInterfaceWithContext(context.Background(), input)
+		prometheusmetrics.Ec2ApiReq.WithLabelValues("CreateNetworkInterface").Inc()
+		prometheusmetrics.AwsAPILatency.WithLabelValues("CreateNetworkInterface", fmt.Sprint(err != nil), awsReqStatus(err)).Observe(msSince(start))
+		if err == nil {
+			log.Infof("Created a new ENI: %s", aws.StringValue(result.NetworkInterface.NetworkInterfaceId))
+			return aws.StringValue(result.NetworkInterface.NetworkInterfaceId), nil
+		}
```

Comment:
This block can just use whatever `ec2.CreateNetworkInterfaceInput` is returned by the custom networking function or subnet search function. That will make this cleaner